### PR TITLE
Fixes #344  refactor generic layer

### LINKED ISSUE #344: Refactor/reengineer Generic

### PR DESCRIPTION
**Pull request by tclune**
## Description
Major restructuring in MAPL

## Motivation and Context
This is for OOMPH (NUOPC Compatibility)

## How Has This Been Tested?
Ran C12 for 1 day.  0 diffs.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Trivial change (affects only documentation or cleanup)

## Checklist:

- [X] I have tested this change with a run of GEOSgcm (if non-trivial)
- [X] I have added one of the required labels (0 diff, 0 diff trivial, 0 diff structural, non 0-diff)
- [X] I have updated the CHANGELOG.md accordingly following the style of [Keep a Changelog](https://keepachangelog.com/en/1.0.0/#how)

### CODE REVIEW

**Comment by tclune** [comment]
Contingent on https://github.com/GEOS-ESM/GMAO_Shared/pull/155

File: CHANGELOG.md [hunk 1]
```diff
@@ -14,6 +14,90 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - For MultiGroupServer, the backend and frontend share each node
 - Moved `MAPL_HistoryGridComp.F90` and related files to `gridcomps/History`
 
+- Major refactoring related to MAPL generic capabilities
+
+    0. Cretaed new subdirectory  "generic"
```
Comment (mathomp4) [inline CHANGELOG.md:19]: Spelling mistake: "Cretaed".

Interestingly, it looks like Markdown does okay with a 0-th led numbered list. Surprising!

File: CHANGELOG.md [hunk 2]
```diff
@@ -14,6 +14,90 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
+       MAPL_MetaComp an extension of AbstractComponentNode, and provided
+       stub implementations for 3 methods. (Commenting out other
+       interfaces in the abstract class for now.)
+    
+    3. Activate get_parent(), add_child(), and num_children()
+    
+    4. Introduced MaplComponentNode
+    
+       Easier to introduce things in a smaller module.   Replaced String NodeMap with
+       an internal map in MaplComponentNdoed wcich uses pointers.   Allows use to
```
Comment (mathomp4) [inline CHANGELOG.md:35]: Spelling: "MaplComponentNdoed" Not sure what this should be...

Comment (tclune) [reply CHANGELOG.md:35]: Yeah - this comment is already out of date.   Closest analog now is ConcreteComposite and StringCompositeMap. Will fix.

File: CHANGELOG.md [hunk 3]
```diff
@@ -14,6 +14,90 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
+       MAPL_MetaComp now just extends MaplComponentNode.
+    
+    5. Extracted internal state.
+    
+    6. Started moving derived types related to import/export specification
+       and such.    The goal will be to then refactor into proper classes
+       with encapsulation.
+    
+    7.  Introducing Vector container for array of pointers to VarpSpecType.
+
```
Comment (atrayano) [inline CHANGELOG.md:46]: Spelling: "VarpSpecType", most likely meant to be "VarSpecType"

**Comment by tclune** [comment]
I removed the Contingent label -  I believe this branch functions just fine.

**Comment by mathomp4** [comment]
@tclune I can confirm the error that CI is getting with GCC 10:
```
[ 61%] Building Fortran object base/CMakeFiles/MAPL.base.dir/MAPL_Generic.F90.o
/discover/swdev/mathomp4/Models/MAPL-GCC-TomBranch/MAPL/base/MAPL_Generic.F90:10279:38:

10279 |        call MAPL_ESMFStateWriteToFile(STATE%get_internal_state(), CLOCK, &
      |                                      1
Error: ‘get_internal_state’ in variable definition context (actual argument to INTENT = OUT/INOUT) at (1) is not a variable
/discover/swdev/mathomp4/Models/MAPL-GCC-TomBranch/MAPL/base/MAPL_Generic.F90:6026:71:

 6026 |        call MAPL_VarSpecSet(SPEC%var_specs%of(L), GRID=GRD, RC=STATUS )
      |                                                                       1
Error: There is no specific subroutine for the generic ‘mapl_varspecset’ at (1)
/discover/swdev/mathomp4/Models/MAPL-GCC-TomBranch/MAPL/base/MAPL_Generic.F90:6202:77:

 6202 |          call MAPL_VarSpecSet(spec%var_specs%of(L),STATE=nestState,RC=STATUS)
      |                                                                             1
Error: There is no specific subroutine for the generic ‘mapl_varspecset’ at (1)
/discover/swdev/mathomp4/Models/MAPL-GCC-TomBranch/MAPL/base/MAPL_Generic.F90:6230:75:

 6230 |          call MAPL_VarSpecSet(SPEC%var_specs%of(L),BUNDLE=BUNDLE,RC=STATUS)
      |                                                                           1
Error: There is no specific subroutine for the generic ‘mapl_varspecset’ at (1)
/discover/swdev/mathomp4/Models/MAPL-GCC-TomBranch/MAPL/base/MAPL_Generic.F90:6431:70:

 6431 |       call MAPL_VarSpecSet(spec%var_specs%of(L),FIELD=FIELD,RC=STATUS)
      |                                                                      1
Error: There is no specific subroutine for the generic ‘mapl_varspecset’ at (1)
base/CMakeFiles/MAPL.base.dir/build.make:302: recipe for target 'base/CMakeFiles/MAPL.base.dir/MAPL_Generic.F90.o' failed
make[2]: *** [base/CMakeFiles/MAPL.base.dir/MAPL_Generic.F90.o] Error 1
CMakeFiles/Makefile2:2122: recipe for target 'base/CMakeFiles/MAPL.base.dir/all' failed
make[1]: *** [base/CMakeFiles/MAPL.base.dir/all] Error 2
Makefile:181: recipe for target 'all' failed
make: *** [all] Error 2
```

**Comment by mathomp4** [comment]
Does this PR still require https://github.com/GEOS-ESM/GMAO_Shared/pull/155 ? I just tried to build the GCM with it (using Intel to avoid the above error) and I'm getting:
```
[100%] Building Fortran object src/Shared/@GMAO_Shared/GEOS_Util/pre/NSIDC-OSTIA_SST-ICE_blend/CMakeFiles/regrid_forcing_esmf.x.dir/regrid_forcing_esmf.F90.o
/discover/swdev/mathomp4/Models/GEOSgcm-Intel-WeiyuanMAPLBranch/GEOSgcm/src/Shared/@GMAO_Shared/GEOS_Util/pre/NSIDC-OSTIA_SST-ICE_blend/regrid_forcing_esmf.F90(582): error #6457: This derived type name has not been declared.   [MAPL_COMMUNICATORS]
  type(MAPL_Communicators) :: mapl_comm
-------^
/discover/swdev/mathomp4/Models/GEOSgcm-Intel-WeiyuanMAPLBranch/GEOSgcm/src/Shared/@GMAO_Shared/GEOS_Util/pre/NSIDC-OSTIA_SST-ICE_blend/regrid_forcing_esmf.F90(606): error #6404: This name does not have a type, and must have an explicit type.   [MAPL_COMM]
  mapl_comm%esmf%comm=MPI_COMM_WORLD
--^
/discover/swdev/mathomp4/Models/GEOSgcm-Intel-WeiyuanMAPLBranch/GEOSgcm/src/Shared/@GMAO_Shared/GEOS_Util/pre/NSIDC-OSTIA_SST-ICE_blend/regrid_forcing_esmf.F90(606): error #6460: This is not a field name that is defined in the encompassing structure.   [ESMF]
  mapl_comm%esmf%comm=MPI_COMM_WORLD
------------^
/discover/swdev/mathomp4/Models/GEOSgcm-Intel-WeiyuanMAPLBranch/GEOSgcm/src/Shared/@GMAO_Shared/GEOS_Util/pre/NSIDC-OSTIA_SST-ICE_blend/regrid_forcing_esmf.F90(606): error #6460: This is not a field name that is defined in the encompassing structure.   [COMM]
  mapl_comm%esmf%comm=MPI_COMM_WORLD
-----------------^
/discover/swdev/mathomp4/Models/GEOSgcm-Intel-WeiyuanMAPLBranch/GEOSgcm/src/Shared/@GMAO_Shared/GEOS_Util/pre/NSIDC-OSTIA_SST-ICE_blend/regrid_forcing_esmf.F90(607): error #6285: There is no matching specific subroutine for this generic subroutine call.   [MAPL_SET]
  call MAPL_Set(MAPLOBJ, mapl_comm = mapl_Comm, rc = status)
-------^
compilation aborted for /discover/swdev/mathomp4/Models/GEOSgcm-Intel-WeiyuanMAPLBranch/GEOSgcm/src/Shared/@GMAO_Shared/GEOS_Util/pre/NSIDC-OSTIA_SST-ICE_blend/regrid_forcing_esmf.F90 (code 1)
```

**Comment by tclune** [comment]
Yes - I got a bit confused at one point.  GMAO_Shared requires changes to work with this MAPL.    I could prevent some leakage but this cruft just needed to go.    So we should first check that the combination works, and then approve in the order:  MAPL, GMAO_Shared, GEOSgcm.    It should be 0-diff at each stage though.

But @bena-nasa  has seen some weird thing that need to be investigated before we approve anyway.

**Comment by tclune** [comment]
The GCC-10 issue looks to be a case of GFortran being more strict than Intel.     I think it is an easy fix, and I'll work on that on Monday.

**Comment by mathomp4** [comment]
FYI to @tclune: I pressed the "Update branch" button

**Comment by tclune** [comment]
It appears that the hang is on most/all of the processes on the second node and at the line:

https://github.com/GEOS-ESM/MAPL/blob/823c173a4d3b7b2baca9c9bf3a0e9c510f7af11d/gridcomps/Cap/MAPL_CapGridComp.F90#L261

Trying to look deeper ...

File: Tests/ExtDataDriverMod.F90 [hunk 4]
```diff
@@ -93,10 +92,9 @@ subroutine run(this,RC)
 
       call this%initialize_io_clients_servers(commCap, rc = status); _VERIFY(status)
       call this%cap_server%get_splitcomm(split_comm)
-      call fill_mapl_comm(split_comm, commCap, this%mapl_comm, rc=status)
       select case(split_comm%get_name())
       case('model')
-         call ESMF_Initialize (vm=vm, logKindFlag=this%cap_options%esmf_logging_mode, mpiCommunicator=this%mapl_comm%esmf%comm, rc=status)
+         call ESMF_Initialize (vm=vm, logKindFlag=this%cap_options%esmf_logging_mode, rc=status)
```
Comment (mathomp4) [inline Tests/ExtDataDriverMod.F90:97]: Does this need a communicator now?

**Comment by tclune** [comment]
probably

I'll add it to the todo list for this branch

**Comment by tclune** [comment]
I just pushed fixes that address the Intel run-time hang as well as unrelated GFortran issues and a separate fix for ExtDataDriver

@bena-nasa Could you retest and be sure to run ExtDataDriver as well?

**Comment by tclune** [comment]
GFortran tests on discover are passing.    We may have to accept this PR with failing CI.

**Comment by mathomp4** [comment]
> GFortran tests on discover are passing. We may have to accept this PR with failing CI.

They pass the tests only because I think you are running `make tests`. The workflow runs:
```yaml
      - name: Run MAPL unit tests
        run: |
          cd build
          make -j4 build-tests
          # skip performance tests
          ctest -R MAPL -LE PERFORMANCE --output-on-failure
```
and it looks like the `generic/tests` does not add to the `build-tests` target. But, `ctest` apparently *does* know about `generic/tests`. So...it complains that `MAPL.generic.tests` was never built. 

I've pushed a commit that moves to use `build-tests`. I tried to figure out how to use `ctest --build-and-test` but I couldn't quite figure it out...

**Comment by mathomp4** [comment]
That last push fixed MAPL, but didn't let GEOS build because of the missing GMAO_Shared. So, in GMAO_Shared, I've "duped" the branch there by pushing a `feature/tclune/#344-refactor-generic` branch to the same hash as `feature/tclune/mapl-generic-refactor`. I've also tried to re-work the mepo in the CI. Let's see if this does it.

**Comment by mathomp4** [comment]
Okay. I think the `checkout-if-exists` works. We'll find out if I just broke it for everything else...

**Review by mathomp4 — APPROVED** [review]
Approving for MAPL 2.6.0

**Pull request merged**